File: php/src/mpc_designation_cli.php
#!/usr/bin/env php
<?php
/**
 * CLI for MPC designation converter
 */

declare(strict_types=1);

require_once __DIR__ . '/MPCDesignation.php';

use MPC\MPCDesignation;
use MPC\MPCDesignationException;

function printUsage(): void {
    $usage = <<<EOF
Usage: php mpc_designation_cli.php [-v|--verbose] <designation> [designation ...]

Convert between packed and unpacked MPC designations.
Auto-detects the input format and converts to the other.

Options:
  -v, --verbose   Show detailed information about the conversion
  --version       Show version information

Examples:
  php mpc_designation_cli.php 00001             -> 1
  php mpc_designation_cli.php 1                 -> 00001
  php mpc_designation_cli.php J95X00A           -> 1995 XA
  php mpc_designation_cli.php '1995 XA'         -> J95X00A
  php mpc_designation_cli.php 'C/1995 O1'       -> CJ95O010
  php mpc_designation_cli.php 1P                -> 0001P
EOF;
    fwrite(STDERR, $usage . "\n");
}

function main(array $argv): int {
    $args = array_slice($argv, 1);

    if (count($args) === 0) {
        printUsage();
        return 1;
    }

    $verbose = false;
    $designations = [];

    foreach ($args as $arg) {
        switch ($arg) {
            case '-v':
            case '--verbose':
                $verbose = true;
                break;
            case '-h':
            case '--help':
                printUsage();
                return 0;
            case '--version':
                echo "mpc_designation " . MPCDesignation::VERSION . "\n";
                return 0;
            default:
                $designations[] = $arg;
                break;
        }
    }

    if (count($designations) === 0) {
        printUsage();
        return 1;
    }

    $multiple = count($designations) > 1;

    foreach ($designations as $des) {
        try {
            $result = MPCDesignation::convert($des);
            $info = $result->info;

            if ($verbose) {
                echo "  Input:    $des\n";
                echo "  Detected: {$info->format} format, {$info->subtype}\n";
                $action = $info->format === 'packed'
                    ? 'unpacking to human-readable form'
                    : 'packing to MPC compact form';
                echo "  Action:   $action\n";
                echo "  Output:   {$result->output}\n";
                if ($multiple) {
                    echo "\n";
                }
            } elseif ($multiple) {
                echo "$des -> {$result->output}\n";
            } else {
                echo "{$result->output}\n";
            }
        } catch (MPCDesignationException $e) {
            fwrite(STDERR, "Error: " . $e->getMessage() . "\n");
            return 1;
        }
    }

    return 0;
}

exit(main($argv));

File: cpp/src/mpc_designation_cli.cpp
/*
 * mpc_designation_cli.cpp - CLI for MPC designation converter
 */

#include "mpc_designation.hpp"
#include <iostream>
#include <vector>
#include <string>

void printUsage() {
    std::cerr << R"(Usage: mpc_designation [-v|--verbose] <designation> [designation ...]

Convert between packed and unpacked MPC designations.
Auto-detects the input format and converts to the other.

Options:
  -v, --verbose   Show detailed information about the conversion
  --version       Show version information

Examples:
  mpc_designation 00001             -> 1
  mpc_designation 1                 -> 00001
  mpc_designation J95X00A           -> 1995 XA
  mpc_designation '1995 XA'         -> J95X00A
  mpc_designation 'C/1995 O1'       -> CJ95O010
  mpc_designation 1P                -> 0001P
)";
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printUsage();
        return 1;
    }

    bool verbose = false;
    std::vector<std::string> designations;

    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (arg == "-h" || arg == "--help") {
            printUsage();
            return 0;
        } else if (arg == "--version") {
            std::cout << "mpc_designation " << mpc::VERSION << "\n";
            return 0;
        } else {
            designations.push_back(arg);
        }
    }

    if (designations.empty()) {
        printUsage();
        return 1;
    }

    bool multiple = designations.size() > 1;

    for (const auto& des : designations) {
        try {
            auto result = mpc::MPCDesignation::convert(des);

            if (verbose) {
                std::cout << "  Input:    " << des << "\n";
                std::cout << "  Detected: "
                          << (result.info.format == mpc::Format::Packed ? "packed" : "unpacked")
                          << " format, " << result.info.subtype << "\n";
                std::string action = (result.info.format == mpc::Format::Packed)
                    ? "unpacking to human-readable form"
                    : "packing to MPC compact form";
                std::cout << "  Action:   " << action << "\n";
                std::cout << "  Output:   " << result.output << "\n";
                if (multiple) {
                    std::cout << "\n";
                }
            } else if (multiple) {
                std::cout << des << " -> " << result.output << "\n";
            } else {
                std::cout << result.output << "\n";
            }
        } catch (const mpc::MPCDesignationError& e) {
            std::cerr << "Error: " << e.what() << "\n";
            return 1;
        }
    }

    return 0;
}
